snr.py

```diff
@@ -0,0 +1,123 @@
+import numpy as np
+import astropy.units as u
+from astropy.tests.helper import assert_quantity_allclose
+
+
+@u.quantity_input(counts=u.electron,
+                  npix=u.pixel,
+                  n_background=u.pixel,
+                  background=u.electron / u.pixel,
+                  darkcurrent=u.electron / u.pixel,
+                  readnoise=u.electron / u.pixel,
+                  gain=u.electron / u.adu,
+                  ad_err=u.adu / u.pixel)
+def snr(counts,
+        npix=1 * u.pixel,
+        n_background=np.inf * u.pixel,
+        background=0 * u.electron / u.pixel,
+        darkcurrent=0 * u.electron / u.pixel,
+        readnoise=0 * u.electron / u.pixel,
+        gain=0 * u.electron / u.adu,
+        ad_err=0.289 * u.adu / u.pixel):
+    """
+    A function to calculate the signal to noise ratio (SNR) of an
+    astronomical observation (from "Handbook of CCD Astronomy",
+    Steve Howell, 2000, pg 55).
+
+    Parameters
+    ----------
+    counts : `~astropy.units.Quantity`
+        Total number of counts in an arbitrary exposure time with units of
+        electrons.
+    npix : `~astropy.units.Quantity`, optional
+        Number of pixels under consideration for the signal with units of
+        pixels. Default is 1 * astropy.units.pixel.
+    n_background : `~astropy.units.Quantity`, optional
+        Number of pixels used in the background estimation with units of
+        pixels. Default is set to np.inf * astropy.units.pixel
+        such that there is no contribution of error due to background
+        estimation. This assumes that n_background will be >> npix.
+    background : `~astropy.units.Quantity`, optional
+        Total photons per pixel due to the backround/sky with units of
+        electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    darkcurrent : `~astropy.units.Quantity`, optional
+        Total electrons per pixel due to the dark current with units
+        of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    readnoise : `~astropy.units.Quantity`, optional
+        Electrons per pixel from the read noise with units of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of electrons/ADU. Default is
+        0 * astropy.units.electron / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        0.289 * astropy.units.adu / astropy.units.pixel
```

Comment:
Ah yeah, the explanation is in a footnote on the next page:
![20190708_160718](https://user-images.githubusercontent.com/33158521/60839260-c265d580-a19a-11e9-90ab-0fe9d1553f41.jpg)
I cite their reference in the docstring if anyone wants to know exactly where it comes from.

Reply:
I see. I have been thinking whether it is more straight forward to ask user to simply provide `gain_err` directly, rather than attempting to calculate it based on assumptions. For instance, HST calculates `gain_err` from actual measurements, not from theory.